compiler-core/src/metadata.rs

```diff
@@ -1,10 +1,234 @@
+                )
+            })
+            .collect::<HashMap<String, compiler_cache::AccessorsHashMap>>();
+
+        let types_constructors = data
+            .types_constructors
+            .iter()
+            .enumerate()
+            .map(|(_, (key, value))| (key.clone(), value.clone()))
+            .collect::<HashMap<String, Vec<String>>>();
+
+        let module = compiler_cache::Module {
+            name: data.name.clone(),
+            types,
+            values,
+            accessors,
+            types_constructors,
+            package: data.package.clone(),
+        };
+
+        let config = bincode::config::standard();
+        let buffer = bincode::serde::encode_to_vec(module, config).expect("bincode");
+
+        Ok(buffer)
+    }
+
+    fn decode_type(id_generator: &UniqueIdGenerator, type_: &Type) -> Type {
+        match (*type_).clone() {
+            Type::App {
+                public,
+                module,
+                name,
+                args,
+            } => Type::App {
+                public: public.clone(),
+                module: module.clone(),
+                name: name.clone(),
+                args: args
+                    .into_iter()
+                    .map(|arg| Arc::new(Metadata::decode_type(&id_generator, &arg)))
+                    .collect::<Vec<Arc<Type>>>(),
+            },
+            Type::Fn { args, retrn } => Type::Fn {
+                args: args
+                    .into_iter()
+                    .map(|arg| Arc::new(Metadata::decode_type(&id_generator, &arg)))
+                    .collect::<Vec<Arc<Type>>>(),
+                retrn: Arc::new(Metadata::decode_type(&id_generator, &retrn)),
+            },
+            Type::Var { type_ } => Type::Var {
+                type_: Arc::new(RefCell::new(match (*type_).clone().into_inner() {
+                    TypeVar::Link { type_: _type } => TypeVar::Link {
+                        type_: Arc::new(Metadata::decode_type(&id_generator, &_type)),
+                    },
+                    TypeVar::Generic { .. } => TypeVar::Generic {
+                        id: id_generator.next(),
+                    },
+                    TypeVar::Unbound { .. } => TypeVar::Unbound {
+                        id: id_generator.next(),
+                    },
+                })),
+            },
+            Type::Tuple { elems } => Type::Tuple {
+                elems: elems
+                    .into_iter()
+                    .map(|elm| Arc::new(Metadata::decode_type(&id_generator, &elm)))
+                    .collect::<Vec<Arc<Type>>>(),
+            },
+        }
+    }
+
+    pub fn decode(
+        id_generator: UniqueIdGenerator,
+        reader: impl BufRead + bincode::de::read::Reader,
+    ) -> Result<Module> {
+        let config = bincode::config::standard();
+        let module: compiler_cache::Module =
+            bincode::serde::decode_from_reader(reader, config).expect("bincode");
+
+        Ok(Module {
```

Comment:
You can avoid all the cloning here by consuming the module when iterating over it rather than taking reference to the values. `into_iter` will help

Reply:
mostly solved now with simplifications